tremor-pipeline/src/op/trickle/select.rs

```diff
@@ -646,221 +648,647 @@ impl Operator for TrickleSelect {
+                                    &env,
+                                    unwind_event,
+                                    state,
+                                    &event_meta,
+                                    &local_stack,
+                                )?;
+                                argv.push(result);
+                            }
+                            for arg in &argv {
+                                argv1.push(arg);
+                            }
+
+                            invocable.accumulate(argv1.as_slice()).map_err(|e| {
+                                // TODO nice error
+                                let r: Option<&Registry> = None;
+                                e.into_err(aggr, aggr, r, &node_meta)
+                            })?;
                         }
+                        false
                     } else {
-                        let s: &Select = &stmt;
-                        return tremor_script::errors::query_guard_not_bool(
-                            s, guard, &test, &node_meta,
-                        )?;
+                        // should not happen
+                        false
+                    };
+
+                    // merge previous aggr state, push window event out, re-initialize for new window
+                    let mut first = true;
+                    let emit_window_iter = emit_window_events.iter().zip(self.windows.iter_mut());
+                    for (window_event, window) in emit_window_iter {
+                        set_window(consts, Value::from(window.name.to_string()))?;
+
+                        // get our group data
+                        let this_groups = &mut window.dims.groups;
+                        let last_groups = &mut window.last_dims.groups;
+                        let window_impl = &window.window_impl;
+                        let mut idgen = self.event_id_gen;
+                        let (_, this_group) = this_groups
+                            .raw_entry_mut()
+                            .from_key(&group_str)
+                            .or_insert_with(|| {
+                                // This is sound since we only add mutability to groups
+                                (
+                                    group_str.clone(),
+                                    last_groups.remove(&group_str).unwrap_or_else(|| GroupData {
+                                        window: window_impl.clone(),
+                                        aggrs: aggregates.clone(),
+                                        group: group_value.clone_static(),
+                                        id: idgen.next_id(), // after all this is a new event
+                                    }),
+                                )
+                            });
+
+                        if window_event.include {
+                            // add this event to the aggr state **BEFORE** emit and propagate to next windows
```

Comment:
I'm a big confused by this, what's scratch?

Reply:
It is a `Vec<InvokeAggrFn>`, a scratch value for the aggregation states.
When we have multiple windows, when the current window emits, we cannot just simply merge the local aggr states into the next window, as this window might need to emit itself before it incorporates the new state. Complicated shizzle. So the only safe way i found is to have basically a tmp value around, where we store the current windows state, before we initialize for the next window. When the next window is ready, it merges the state from the scratch value.
We only ever need max 2 scratch "registers" for the whole select statement execution at any time, i am thinking about putting it into the `SelectStatement` struct, but wanted to ensure my approach flies first.

Reply:
Ah I think I get it. Let me try to paraphrase. We clone the value we want to merge so we can reinitialize the old window (kill it's state) emit the pre-merge state of the next without?